std/var: support floating point correction value

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94073
* #94812

Ref https://github.com/pytorch/pytorch/issues/61492#issuecomment-1413003480

The array API specifies correction to be `Union[int, float]` while we currently only support integers.
https://data-apis.org/array-api/latest/API_specification/generated/array_api.std.html

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

peterbell10 committed:
std/var: support floating point correction value

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


Ref https://github.com/pytorch/pytorch/issues/61492#issuecomment-1413003480

The array API specifies correction to be `Union[int, float]` while we currently only support integers.
https://data-apis.org/array-api/latest/API_specification/generated/array_api.std.html

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ezyang commented on torchgen/dest/lazy_ts_lowering.py:
```diff
@@ -1,28 +1,28 @@
     # and we use the functional version not out/inplace.
     emplace_arguments = []
+
+    def get_value(arg: LazyArgument) -> str:
```
Hmm what's going on here?

peterbell10 replied:
Lazy tensor IR doesn't store operands which are `nullopt`, but the logic to skip reading the operand was not applied to keyword-only arguments resulting in an out of bounds vector access.

This PR runs into this bug because `int?` is stored as a property on the IR node and not as an operand, whereas `Scalar?` is an operand. You can see the test failure I was getting below.

```
======================================================================
ERROR [0.004s]: test_correctness_std_cpu_float32 (__main__.TestLazyOpInfoCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 414, in instantiated_test
    raise rte
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 401, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 851, in test_wrapper
    return test(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/lazy/test_ts_opinfo.py", line 241, in test_correctness
    assert_allclose_rec((r_actual, r_exp))
  File "/var/lib/jenkins/workspace/test/lazy/test_ts_opinfo.py", line 227, in assert_allclose_rec
    self.assertTrue(torch.allclose(clone_to_device(a, test_device), b, atol=1e-4))
  File "/var/lib/jenkins/workspace/test/lazy/test_ts_opinfo.py", line 218, in clone_to_device
    return input.detach().clone().to(device=dev)
RuntimeError: vector::_M_range_check: __n (which is 1) >= this->size() (which is 1)
```

ezyang reviewed (approved):
okey dokey

ezyang commented:
@pytorchbot merge -r master

peterbell10 committed:
Update on "std/var: support floating point correction value"


Ref https://github.com/pytorch/pytorch/issues/61492#issuecomment-1413003480

The array API specifies correction to be `Union[int, float]` while we currently only support integers.
https://data-apis.org/array-api/latest/API_specification/generated/array_api.std.html

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


Ref https://github.com/pytorch/pytorch/issues/61492#issuecomment-1413003480

The array API specifies correction to be `Union[int, float]` while we currently only support integers.
https://data-apis.org/array-api/latest/API_specification/generated/array_api.std.html

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


Ref https://github.com/pytorch/pytorch/issues/61492#issuecomment-1413003480

The array API specifies correction to be `Union[int, float]` while we currently only support integers.
https://data-apis.org/array-api/latest/API_specification/generated/array_api.std.html

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


Ref https://github.com/pytorch/pytorch/issues/61492#issuecomment-1413003480

The array API specifies correction to be `Union[int, float]` while we currently only support integers.
https://data-apis.org/array-api/latest/API_specification/generated/array_api.std.html

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "std/var: support floating point correction value"


Ref https://github.com/pytorch/pytorch/issues/61492#issuecomment-1413003480

The array API specifies correction to be `Union[int, float]` while we currently only support integers.
https://data-apis.org/array-api/latest/API_specification/generated/array_api.std.html

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge -f "CI is green, stacked PR failure is unrelated flake"

peterbell10 committed:
Update on "std/var: support floating point correction value"


Ref https://github.com/pytorch/pytorch/issues/61492#issuecomment-1413003480

The array API specifies correction to be `Union[int, float]` while we currently only support integers.
https://data-apis.org/array-api/latest/API_specification/generated/array_api.std.html

As std/var is calculated currently, the final count of elements is already done
in floating point so we can make the correction floating point without any loss
of precision or generality.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
